src/libxively/sft/xi_sft_logic.c

```diff
@@ -62,37 +68,178 @@ xi_state_t xi_sft_on_connected( xi_sft_context_t* context )
-               - get the first chunk of file in question with FILE_GET_CHUNK */
-            break;
-        case XI_CONTROL_MESSAGE_SC_FILE_CHUNK:
+        case XI_CONTROL_MESSAGE_SC__SFT_FILE_UPDATE_AVAILABLE:
+        {
+            /* todo?: check whether device is ready to start download of file */
+
+            if ( NULL != context->update_message_fua )
+            {
+                xi_control_message_free( &context->update_message_fua );
+            }
+
+            context->update_message_fua = sft_message_in;
+            /* prevent deallocation */
+            sft_message_in = NULL;
+
+            if ( 0 < context->update_message_fua->file_update_available.list_len &&
+                 NULL != context->update_message_fua->file_update_available.list )
+            {
+                context->update_current_file =
+                    &context->update_message_fua->file_update_available.list[0];
+
+                xi_control_message_t* message_file_get_chunk =
+                    xi_control_message_create_file_get_chunk(
+                        context->update_current_file->name,
+                        context->update_current_file->revision, 0,
+                        XI_MIN( XI_SFT_DOWNLOAD_BYTES_PER_FILE_CHUNK,
+                                context->update_current_file->size_in_bytes ) );
+
+                ( *context->fn_send_message )( context->send_message_user_data,
+                                               message_file_get_chunk );
+            }
+        }
+        break;
+        case XI_CONTROL_MESSAGE_SC__SFT_FILE_CHUNK:
+        {
             /* - store the chunk through file/firmware BSP
-               - get the next chunk, XI_CONTROL_MESSAGE_CS_FILE_GET_CHUNK
                - if last chunk close file, start appropriate action
-                    - if firmware: start firmware update or "just" report downloaded and
+                    - if firmware: start firmware update or "just" report downloaded
+               and
                       wait for remote FWU trigger
-                    - other files may have their own action, no-operation is an option */
-            break;
+                    - other files may have their own action, no-operation is an option
+               */
+
+            // todo: check whether FILE_CHUNK belongs to exactly to which file was
+            // requested
+
+            if ( NULL != context->update_current_file &&
+                 NULL != context->update_current_file->name &&
+                 NULL != sft_message_in->file_chunk.name &&
+                 0 == strcmp( context->update_current_file->name,
+                              sft_message_in->file_chunk.name ) )
+            {
+                const uint32_t all_downloaded_bytes =
+                    sft_message_in->file_chunk.offset + sft_message_in->file_chunk.length;
+
+                printf( "         === === === downloading file: %s, %d / %d, [%d%%]\n",
```

Comment:
I'll remove it later but this is too fresh it is good to see the download progress